### DISCUSSION TITLE
rdflib.term.fromn3 method

### DISCUSSION THREAD

**Discussion by ghost** [RDFLib/rdflib#1532] [Ideas]
gromgull, 2012-01-25T10:46:53.000Z

RDFLib Terms have a .n3() method that return the n3 representation of the term. 

For round-tripping, it would be nice to have a method that does the opposite, i.e. something that creates a term object from the n3 representation.
### Comment 1 by eik...@eikeon.com

see: 

``` python
from rdflib.util import from_n3
```
### Comment 2 by joernhees

thanks. would still be cool to have that in term.
even cooler would be if the n3 parser used it in a similar way as the n3 serializer uses .n3()

**Comment by gweis** [comment]
As far as I can tell, the TSV Parser could use it as well.
However from_n3 doesn't to much validation. I guess it would be good to add validation to from_n3, so that it could be re-used in all parsers.

**Comment by gromgull** [comment]
I would think it should be possible to have a common code base for parsing rdf1.1 ntriples (utf8!) and turtle/n3 nodes.

**Comment by joernhees** [comment]
related: #502, #503

as mentioned i also think it would be useful to only have one place where we parse n3... #503 fixes `util.from_n3()`'s parsing of literals with datatypes, and allows parsing PrefixedNames by introducing a `NamespaceManager`.

Still, as `util.from_n3()` is a "single term parser", it for example doesn't really seem to do well with parsing quoted graphs.

I think some of the points to decide here are:
- [ ] should we unify the parsing of n3?
  - [ ] should we use `parser/n3.py` code from `util.from_n3()` or
  - [ ] vice versa?
- [ ] should we be able to parse all `rdflib.term`s from n3, e.g. `Variable`, `QuotedGraph`?
